pubsub.go

```diff
@@ -397,13 +433,69 @@ func (p *PubSub) processLoop(ctx context.Context) {
 				p.rt.RemovePeer(pid)
 			}
 
+		case discover := <-p.rediscover:
+			topic := discover.topic
+			if _, ok := p.ongoingDiscovery[topic]; !ok {
+				p.ongoingDiscovery[topic] = struct{}{}
+				go func() {
+					p.handleDiscovery(p.ctx, topic, discover.opts)
+					p.doneRediscovery <- topic
+				}()
+			}
+		case topic := <-p.doneRediscovery:
+			delete(p.ongoingDiscovery, topic)
+
 		case <-ctx.Done():
 			log.Info("pubsub processloop shutting down")
 			return
 		}
 	}
 }
 
+func (p *PubSub) handleDiscovery(ctx context.Context, topic string, opts []discovery.Option) {
+	if p.discovery == nil {
+		return
+	}
+
+	discoverCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	peerChan, err := p.discovery.FindPeers(discoverCtx, topic, opts...)
+	if err != nil {
+		log.Debugf("error finding peers for topic %s: %v", topic, err)
+		return
+	}
+
+	for {
+		select {
+		case pi, more := <-peerChan:
+			go func() {
+				ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+				defer cancel()
+
+				err := p.host.Connect(ctx, pi)
+				if err != nil {
+					log.Debugf("Error connecting to pubsub peer %s: %s", pi.ID, err.Error())
+					return
+				}
+
+				// delay to let pubsub perform its handshake
+				time.Sleep(time.Millisecond * 250)
```

Comment:
We shouldn't need to wait here (it's not like we're doing anything but printing a debug line after).

Reply:
Sleeping is never a good option. There is no guarantee that 250ms is enough.
I understand that this is an old code but it probably should be solved in another way.

Reply:
Fair enough, now that this code is in pubsub instead of in go-libp2p-pubsub-router we should have enough info to not need to sleep

Reply:
The difference is the lack of a waitgroup in this case.

Reply:
yep, that's my bad. If I'm not going to remove the delay then I need to add the waitgroup back in. Any preferences @Stebalien